lib/matplotlib/_type1font.py

```diff
@@ -768,13 +861,262 @@ def transform(self, effects):
+                    16: 'callothersubr',
+                    17: 'pop',
+                    33: 'setcurrentpoint'
+                }[byte1]
+            else:
+                yield {
+                    1: 'hstem',
+                    3: 'vstem',
+                    4: 'vmoveto',
+                    5: 'rlineto',
+                    6: 'hlineto',
+                    7: 'vlineto',
+                    8: 'rrcurveto',
+                    9: 'closepath',
+                    10: 'callsubr',
+                    11: 'return',
+                    13: 'hsbw',
+                    14: 'endchar',
+                    21: 'rmoveto',
+                    22: 'hmoveto',
+                    30: 'vhcurveto',
+                    31: 'hvcurveto'
+                }[byte]
+
+    def _simulate(
+        self, glyph_or_subr: str | int, buildchar_stack: list[float],
+        postscript_stack: list[float]
+    ) -> tuple[set[str], set[int], list[float], list[float]]:
+        """Run the charstring interpreter on a glyph or subroutine.
+
+        This does not actually execute the code but simulates it to find out
+        which subroutines get called when executing the glyph or subroutine.
+
+        Returns
+        -------
+            glyphs : set[str]
+                The set of glyph names that are called by the glyph or subroutine
+            subrs : set[int]
+                The set of subroutines that are called by the glyph or subroutine
+            buildchar_stack : list[float]
+                The buildchar stack at the end of the simulation
+            postscript_stack : list[float]
+                The PostScript stack at the end of the simulation
+        """
+        if isinstance(glyph_or_subr, str):
+            program = self.prop['CharStrings'][glyph_or_subr]
+            glyphs = {glyph_or_subr}
+            subrs = set()
+        else:
+            program = self.prop['Subrs'][glyph_or_subr]
+            glyphs = set()
+            subrs = {glyph_or_subr}
+        for opcode in self._charstring_tokens(program):
+            if opcode in ('return', 'endchar'):
+                return glyphs, subrs, buildchar_stack, postscript_stack
+            newglyphs, newsubrs, buildchar_stack, postscript_stack = \
+                self._step(buildchar_stack, postscript_stack, opcode)
+            glyphs.update(newglyphs)
+            subrs.update(newsubrs)
+        else:
+            font_name = self.prop.get('FontName', '(unknown)')
+            _log.info(
+                f"Glyph or subr {glyph_or_subr} in font {font_name} does not end "
+                "with return or endchar"
+            )
+            return glyphs, subrs, buildchar_stack, postscript_stack
+
+    def _step(
+            self,
+            buildchar_stack: list[float],
+            postscript_stack: list[float],
+            opcode: int | str,
+        ) -> tuple[set, set, list[float], list[float]]:
+        """Run one step in the charstring interpreter."""
```

Comment:
I wonder if this may be clearer if you mutate buildchar_stack and postscript_stack in-place?  this way you would write
```python
glyphs = set(); subrs = set()
if opcode in {...}:
    buildchar_stack[:] = []
elif opcode == "seac":
    codes = ...; glyphs.update(...)
    buildchar_stack[:] = []
elif opcode == "div":
    num2 = buildchar_stack.pop()
    num1 = buildchar_stack.pop()
    buildchar_stack.append(num1 / num2)
...
return glyphs, subrs
```
which feels perhaps more in the spirit of a postscript interpreter?

Reply:
Could be! I'll think about this a little.

Reply:
I moved this into a separate class with the stacks and glyph/subr sets as members. I agree that it looks clearer that way.